src/abstractdataframe/selectionfast.jl

```diff
@@ -49,15 +53,91 @@ julia> sum(Any[1.0, typemax(Int), typemax(Int)])
+"""
+isreduction(::Any) = false
+isreduction(::typeof(sum)) = true
+isreduction(::typeof(sum∘skipmissing)) = true
+isreduction(::typeof(length)) = true
+isreduction(::typeof(mean)) = true
+isreduction(::typeof(mean∘skipmissing)) = true
+isreduction(::typeof(var)) = true
+isreduction(::typeof(var∘skipmissing)) = true
+isreduction(::typeof(std)) = true
+isreduction(::typeof(std∘skipmissing)) = true
+isreduction(::typeof(median)) = true
+isreduction(::typeof(median∘skipmissing)) = true
+isreduction(::typeof(minimum)) = true
+isreduction(::typeof(minimum∘skipmissing)) = true
+isreduction(::typeof(maximum)) = true
+isreduction(::typeof(maximum∘skipmissing)) = true
+isreduction(::typeof(prod)) = true
+isreduction(::typeof(prod∘skipmissing)) = true
+isreduction(::typeof(first)) = true
+isreduction(::typeof(first∘skipmissing)) = true
+isreduction(::typeof(last)) = true
+
 """
     default_table_transformation(df_sel::AbstractDataFrame, fun)
 
 This is a default implementation called when `AsTable(...) => fun` is requested.
 The `df_sel` argument is a data frame storing columns selected by
 `AsTable(...)` selector.
 """
-default_table_transformation(df_sel::AbstractDataFrame, fun) =
-    fun(Tables.columntable(df_sel))
+function default_table_transformation(df_sel::AbstractDataFrame, fun)
+    if fun isa ByRow && fun.fun isa ComposedFunction{<:Any, typeof(collect)}
+        vT = unique(typeof.(eachcol(df_sel)))
+        if length(vT) == 1 # homogeneous type
+            T = eltype(vT[1])
+            cT = vT[1]
+        elseif length(vT) == 2 # small union
+            # Base.promote_typejoin is used wen collecting NamedTuple elements
+            T = Base.promote_typejoin(eltype(vT[1]), eltype(vT[2]))
+            cT = Union{vT[1], vT[2]}
+        else # large union
+            # use Base.promote_typejoin to make sure that in case all columns
+            # have <:Real eltype the v vector has <:Real eltype which
+            # is required by some functions, eg. in StatsBase.jl
+            T = mapreduce(eltype, Base.promote_typejoin, vT)
+            # use Any for cT, as the common type of columns will be abstract
+            # anyway, so it is better to use Any as it should reduce compilation latency
+            cT = Any
+        end
+        v = Vector{T}(undef, ncol(df_sel))
+        cols = collect(cT, eachcol(df_sel))
+        reduction = isreduction(fun.fun.outer)
+        return _fast_row_aggregate_collect(fun.fun.outer, v, cols, reduction)
+    elseif fun isa ComposedFunction{<:Any, typeof(collect)}
+        # this will narrow down eltype of the resulting vector
+        # but will not perform conversion
+        return fun(map(identity, eachcol(df_sel)))
+    else
+        return fun(Tables.columntable(df_sel))
+    end
+end
+
+function _populate_v!(v::Vector, cols::Vector, len::Int, i::Int, reduction::Bool)
+    for j in 1:len
+        @inbounds v[j] = cols[j][i]
+    end
+    return reduction ? v : copy(v)
```

Comment:
@nalimilan following your last comment I have made also the following change. I make sure to reuse the working vector for rows only for functions that are reductions (i.e. are guaranteed not to reuse or resize the collected vectors). This makes sure that the following is correct:
```
julia> df = DataFrame(rand(10, 5), :auto)
10×5 DataFrame
 Row │ x1         x2         x3         x4         x5       
     │ Float64    Float64    Float64    Float64    Float64  
─────┼──────────────────────────────────────────────────────
   1 │ 0.569168   0.441972   0.134152   0.0920753  0.935849
   2 │ 0.338254   0.917705   0.447445   0.313503   0.41281
   3 │ 0.0164164  0.936971   0.791789   0.678257   0.551078
   4 │ 0.685475   0.731142   0.405247   0.715096   0.27323
   5 │ 0.658448   0.885822   0.541997   0.796662   0.331429
   6 │ 0.679871   0.806971   0.574006   0.700705   0.374933
   7 │ 0.964579   0.873352   0.485584   0.729399   0.963421
   8 │ 0.0093121  0.478524   0.0092773  0.217818   0.928038
   9 │ 0.124796   0.0379127  0.139757   0.0520611  0.49285
  10 │ 0.0715453  0.275452   0.197599   0.212332   0.930255

julia> combine(df, AsTable(All()) => ByRow(identity∘collect))
10×1 DataFrame
 Row │ x1_x2_etc_identity_collect        
     │ Vector{Float64}
─────┼───────────────────────────────────
   1 │ [0.569168, 0.441972, 0.134152, 0…
   2 │ [0.338254, 0.917705, 0.447445, 0…
   3 │ [0.0164164, 0.936971, 0.791789, …
   4 │ [0.685475, 0.731142, 0.405247, 0…
   5 │ [0.658448, 0.885822, 0.541997, 0…
   6 │ [0.679871, 0.806971, 0.574006, 0…
   7 │ [0.964579, 0.873352, 0.485584, 0…
   8 │ [0.0093121, 0.478524, 0.0092773,…
   9 │ [0.124796, 0.0379127, 0.139757, …
  10 │ [0.0715453, 0.275452, 0.197599, …
```
at the expense of additional copying when the operation is not a reduction.

Reply:
Actually I had also just realized the same issue. :-) Indeed it sounds safer to restrict to some functions known as safe. 

Though `isreduction` is a bit too specific, right? This is actually more something like `isnonmutating` AFAICT. (This is again a case where being able to declare that a function is pure would be useful.)

Reply:
No - this is more than pure. We require some kind of reduction.
The point is that
not only we are not allowed to resize the passed argument
but also we are not allowed to return it in any form (pure function could do this).

For example `identity` is pure, but as a result (without my current fix) it would produce the same vector in all rows (it would contain the last row of the source table in my example above).

This of course extends not only to just returning a vector, but also wrapping it, making a view, etc. - all this disallowed.

Reply:
Ah right. Though functions which are not reductions at all are still fine as they make copies, e.g. `copy`, `normalize`, `sort`... Not sure how to call this. `iscopying`? `isaliased`? `isinplace`?

Reply:
> Though functions which are not reductions at all are still fine as they make copies

Yes

---

Maybe `is_reading_only` as this is essentially what we require - that the passed vector is only used for reading its contents?

Anyway - I have made this function not part of a public API for now.

Reply:
Yeah, or maybe just `isreadonly`?

Reply:
OK - changed.